pyrokinetics/equilibrium.py

```diff
@@ -159,6 +162,173 @@ def read_geqdsk(self, psi_n_lcfs=1.0):
+            try:
+                R_mom_cos[i, :, :] = (
+                    np.cos(i * theta) * data[f"RMC{i:02d}"][itime, :] * 1e-2
+                )
+            except IndexError:
+                break
+            Z_mom_cos[i, :, :] = (
+                np.cos(i * theta) * data[f"YMC{i:02d}"][itime, :] * 1e-2
+            )
+            if i == 0:
+                R_mom_sin[i, :, :] = 0.0
+                Z_mom_sin[i, :, :] = 0.0
+            else:
+                R_mom_sin[i, :, :] = (
+                    np.sin(i * theta) * data[f"RMS{i:02d}"][itime, :] * 1e-2
+                )
+                Z_mom_sin[i, :, :] = (
+                    np.sin(i * theta) * data[f"YMS{i:02d}"][itime, :] * 1e-2
+                )
+
+        Rsur = np.sum(R_mom_cos, axis=0) + np.sum(R_mom_sin, axis=0)
+        Zsur = np.sum(Z_mom_cos, axis=0) + np.sum(Z_mom_sin, axis=0)
+
+        psi_axis = data["PSI0_TR"][itime]
+        psi_bdry = data["PLFLXA"][itime]
+
+        current = data["PCUR"][itime]
+
+        # Load in 1D profiles
+        q = data["Q"][itime, :]
+        press = data["PMHD_IN"][itime, :]
+
+        # F is on a different grid and need to remove duplicated HFS points
+        psi_rmajm = data["PLFMP"][itime, :]
+        rmajm_ax = np.argmin(psi_rmajm)
+        psi_n_rmajm = psi_rmajm[rmajm_ax:] / psi_rmajm[-1]
+
+        # f = (Bt / |B|) * |B| *  R
+        f = (
+            data["FBTX"][itime, rmajm_ax:]
+            * data["BTX"][itime, rmajm_ax:]
+            * data["RMAJM"][itime, rmajm_ax:]
+            * 1e-2
+        )
+
+        psi = data["PLFLX"][itime, :]
+        psi_n = psi / psi[-1]
+
+        rbdry = Rsur[:, -1]
+        zbdry = Zsur[:, -1]
+
+        Rmin = min(rbdry)
+        Rmax = max(rbdry)
+        Zmin = min(zbdry)
+        Zmax = max(zbdry)
+
+        Rgrid = np.linspace(Rmin, Rmax, nr)
+        Zgrid = np.linspace(Zmin, Zmax, nr)
+
+        # Set up 1D profiles
+
+        # Using interpolated splines
+        q_interp = InterpolatedUnivariateSpline(psi_n, q)
+        press_interp = InterpolatedUnivariateSpline(psi_n, press)
+        f_interp = InterpolatedUnivariateSpline(psi_n_rmajm, f)
+        f2_interp = InterpolatedUnivariateSpline(psi_n_rmajm, f ** 2)
+        ffprime_interp = f2_interp.derivative()
+
+        # Set up 2D profiles
+        # Re-map from R(theta, psi), Z (theta, psi) to psi(R, Z)
+        Rmesh, Zmesh = np.meshgrid(Rgrid, Zgrid)
+        Rmesh_flat = np.ravel(Rmesh)
+        Zmesh_flat = np.ravel(Zmesh)
+
+        Rflat = np.ravel(Rsur.T)
+        Zflat = np.ravel(Zsur.T)
+        psiflat = np.repeat(psi, ntheta)
+
+        RZflat = np.stack([Rflat, Zflat], -1)
+        RZmesh_flat = np.stack([Rmesh_flat, Zmesh_flat], -1)
+
+        # Interpolate using flat data
+        psiRZ_interp = RBFInterpolator(RZflat, psiflat, kernel="cubic")
+
+        # Map data to new grid and reshape
+        psiRZ_data = np.reshape(psiRZ_interp(RZmesh_flat), np.shape(Rmesh)).T
```

Comment:
What does this do for the psi outside the LCFS? Just wondering if it's possible that this could create "phantom" contours outside the LCFS when looking for flux surfaces.

Reply:
It does allow for extrapolation so you could do something like this

```python
    # Interpolate using flat data
    psiRZ_interp = RBFInterpolator(RZflat, psiflat, kernel='cubic')

    # Set up new grid with slightly larger domain
    Rgrid = np.linspace(Rmin*0.9, Rmax*1.1, gq_nr)
    Zgrid = np.linspace(Zmin*1.1, Zmax*1.1, gq_nz)

    rdim = max(Rgrid) - min(Rgrid)
    zdim = max(Zgrid) - min(Zgrid)
    rleft = min(Rgrid)
    zmid = (max(Zgrid) + min(Zgrid)) / 2
    nx = gq_nr
    ny = gq_nz

    Rmesh, Zmesh = np.meshgrid(Rgrid, Zgrid)
    Rmesh_flat = np.ravel(Rmesh)
    Zmesh_flat = np.ravel(Zmesh)
    RZmesh_flat = np.stack([Rmesh_flat, Zmesh_flat], -1)
    
    # Map data to new grid and reshape
    psiRZ_data = np.reshape(psiRZ_interp(RZmesh_flat), np.shape(Rmesh)).T
```

Works quite nicely

Reply:
Added options in https://github.com/pyro-kinetics/pyrokinetics/pull/48/commits/39f1d67f70d0d8b4190c5b2e90b9de012c76c1bf